### ISSUE TITLE
Problem with integrating blaze MoEGate

### ISSUE THREAD

**Issue by ronnie00TT** [tenstorrent/tt-metal#42198]
A new MoEGate has been integrated in this [branch](https://github.com/tenstorrent/tt-metal/tree/ronnie/add_new_moe_gate_rc1_latest). And the accuracy is very high in test_moe_gate.py using real weights (weights pcc>0.999, indices correctness>0.8, both higher than the previous MoEGate).
And now something strange is observed in `test_demo.py` using `dual_full_demo_32upr` test case:

1. If trace_mode is ON   and sampling on device is OFF, the result is CORRECT.
2. If trace_mode is OFF and sampling on device is ON  , the result is CORRECT.
3. If trace_mode is OFF and sampling on device is OFF , the result is CORRECT.
4. If trace_mode is ON.  and sampling on device is ON  , the result is WRONG (including characters from diderent languages).

Also, the new MoEGate has been tested in trace mode using test_moe_gate.py. And the output is exactly the same as non-trace mode.
And on `ds-rc1` branch, the output is correct even with trace_mode and sampling on device ON

**Comment by pprajapatiTT** [comment]
since` trace off + sampling on device` works well and `trace on + sampling on device` produces bad o/p, this seems like more of synchronization issue..


Try this in `models/demos/deepseek_v3/tt/generator.py#2029` (non-MTP path) decode loop.

1.  add synchronize_device
```
ttnn.synchronize_device(self.mesh_device)
          if self.sample_on_device:
              pred_tokens_device = self._sample_tokens_device(
                  decode_logits, enable_trace=self.enable_trace
              )
```

2. synchronize_device and turn off tracing of sampling module
```
ttnn.synchronize_device(self.mesh_device)
          if self.sample_on_device:
              pred_tokens_device = self._sample_tokens_device(
                  decode_logits, enable_trace=Fasle
              )
```

**Comment by ronnie00TT** [comment]
Thanks @pprajapatiTT , I ran above tests, but the results are still wrong.

**Comment by pprajapatiTT** [comment]
Next step is to compare logits after decode step and before the sampling..

So, 
loop {
  device_logits =  decode_forward()
   <capture device_logits>
   sampling-on-device()
}

compare the logits for these two cases..
1. trace off  + sampling on device
2. trace on  + sampling on device

if logits are same in both cases then something wrong with sampling, else some issue with decode_forward()

**Comment by ronnie00TT** [comment]
Thanks @pprajapatiTT , I think the problem might be with sampling. SInce I found that half of the `device_logits` are correct for the first user, and it keeping decreasing for the following users.

**Comment by pprajapatiTT** [comment]
If sampling module is getting only half of the logits correct then sampling will, ofcourse, produce garbage. That's not issue with sampling, that's issue with layer above it who gives inputs to sampling.
Let's have a debug session...

**Comment by ronnie00TT** [comment]
After a further debug (on a single galaxy), I found that there is rand function inside the generator. And even if I keep using the same test case, the outputs are different.

**Comment by pprajapatiTT** [comment]
`ed8a9962268b9864dd037be0ccb21ac5d8ef21c2` commit implements the warmup path in demo. With that buffer corruptions are avoided and bad o/p issue mentioned in this ticket is resolved.

**Comment by pprajapatiTT** [comment]
warmup flow is merged in ds-rc1, that fixes this issue. closing
